File: mutate4go/exec.go
package mutate4go

import (
	"context"
	"os/exec"
	"time"
)

func (e ProcessCommandExecutor) Run(command []string, dir string, timeoutMillis int64) (CommandResult, error) {
	var (
		ctx    context.Context
		cancel context.CancelFunc
	)
	if timeoutMillis > 0 {
		ctx, cancel = context.WithTimeout(context.Background(), time.Duration(timeoutMillis)*time.Millisecond)
	} else {
		ctx, cancel = context.WithCancel(context.Background())
	}
	defer cancel()

	started := time.Now()
	cmd := exec.CommandContext(ctx, command[0], command[1:]...)
	cmd.Dir = dir
	output, err := cmd.CombinedOutput()
	duration := time.Since(started).Milliseconds()
	if ctx.Err() == context.DeadlineExceeded {
		return CommandResult{ExitCode: 124, Output: string(output), DurationMillis: duration, TimedOut: true}, nil
	}
	if err == nil {
		return CommandResult{ExitCode: 0, Output: string(output), DurationMillis: duration}, nil
	}
	if exitErr, ok := err.(*exec.ExitError); ok {
		return CommandResult{ExitCode: exitErr.ExitCode(), Output: string(output), DurationMillis: duration}, nil
	}
	return CommandResult{}, err
}

func NewProcessTestCommandExecutor(command []string) ProcessTestCommandExecutor {
	return ProcessTestCommandExecutor{command: command, runner: ProcessCommandExecutor{}}
}

func (e ProcessTestCommandExecutor) WithCommand(command string) ProcessTestCommandExecutor {
	e.shellCommand = command
	return e
}

func (e ProcessTestCommandExecutor) RunTests(projectRoot string, timeoutMillis int64) (TestRun, error) {
	command := e.command
	if len(command) == 0 {
		command = []string{"go", "test", "./..."}
	}
	if e.shellCommand != "" {
		command = []string{"sh", "-lc", e.shellCommand}
	}
	result, err := e.runner.Run(command, projectRoot, timeoutMillis)
	if err != nil {
		return TestRun{}, err
	}
	return TestRun{
		ExitCode:       result.ExitCode,
		Output:         result.Output,
		DurationMillis: result.DurationMillis,
		TimedOut:       result.TimedOut,
	}, nil
}

File: mutate4rs/src/exec.rs
use std::{
    path::Path,
    process::{Command, Stdio},
    thread,
    time::{Duration, Instant},
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CommandResult {
    pub exit_code: i32,
    pub output: String,
    pub duration_millis: u128,
    pub timed_out: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TestRun {
    pub exit_code: i32,
    pub output: String,
    pub duration_millis: u128,
    pub timed_out: bool,
}

pub struct ProcessCommandExecutor;

impl ProcessCommandExecutor {
    pub fn run(
        &self,
        command: &[&str],
        working_directory: &Path,
        timeout_millis: u64,
    ) -> Result<CommandResult, String> {
        let Some(program) = command.first() else {
            return Err("command must not be empty".to_string());
        };
        let start = Instant::now();
        let mut child = Command::new(program)
            .args(&command[1..])
            .current_dir(working_directory)
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()
            .map_err(|err| err.to_string())?;

        if timeout_millis > 0 {
            let timeout = Duration::from_millis(timeout_millis);
            loop {
                if let Some(_status) = child.try_wait().map_err(|err| err.to_string())? {
                    let output = child.wait_with_output().map_err(|err| err.to_string())?;
                    return Ok(CommandResult {
                        exit_code: output.status.code().unwrap_or(1),
                        output: String::from_utf8_lossy(&output.stdout).to_string()
                            + &String::from_utf8_lossy(&output.stderr),
                        duration_millis: start.elapsed().as_millis(),
                        timed_out: false,
                    });
                }
                if start.elapsed() >= timeout {
                    let _ = child.kill();
                    let output = child.wait_with_output().map_err(|err| err.to_string())?;
                    return Ok(CommandResult {
                        exit_code: 124,
                        output: String::from_utf8_lossy(&output.stdout).to_string()
                            + &String::from_utf8_lossy(&output.stderr),
                        duration_millis: start.elapsed().as_millis(),
                        timed_out: true,
                    });
                }
                thread::sleep(Duration::from_millis(5));
            }
        }

        let output = child.wait_with_output().map_err(|err| err.to_string())?;
        Ok(CommandResult {
            exit_code: output.status.code().unwrap_or(1),
            output: String::from_utf8_lossy(&output.stdout).to_string()
                + &String::from_utf8_lossy(&output.stderr),
            duration_millis: start.elapsed().as_millis(),
            timed_out: false,
        })
    }
}

#[derive(Clone)]
pub struct ProcessTestCommandExecutor {
    command: Vec<String>,
    shell_command: Option<String>,
}

impl ProcessTestCommandExecutor {
    pub fn new(command: Vec<String>) -> Self {
        Self {
            command,
            shell_command: None,
        }
    }

    pub fn with_command(mut self, command: &str) -> Self {
        self.shell_command = Some(command.to_string());
        self
    }

    pub fn run_tests(&self, project_root: &Path, timeout_millis: u64) -> Result<TestRun, String> {
        let runner = ProcessCommandExecutor;
        let result = if let Some(command) = &self.shell_command {
            runner.run(&["sh", "-lc", command], project_root, timeout_millis)?
        } else if self.command.is_empty() {
            runner.run(&["cargo", "test"], project_root, timeout_millis)?
        } else {
            let refs = self.command.iter().map(String::as_str).collect::<Vec<_>>();
            runner.run(&refs, project_root, timeout_millis)?
        };

        Ok(TestRun {
            exit_code: result.exit_code,
            output: result.output,
            duration_millis: result.duration_millis,
            timed_out: result.timed_out,
        })
    }
}
